### PR TITLE
Factor out machinery for running subprocess tk tests.

### PR DESCRIPTION
**Pull request by anntzer**
Probably it can also be shared with test_interactive_backends in the
future.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_tk.py [hunk 1]
```diff
@@ -8,147 +11,121 @@
...
 # NOTE: TkAgg tests seem to have interactions between tests,
 # So isolate each test in a subprocess. See GH#18261
+# Each of the tests decorated with _tk_test must be fully self-contained, and
+# thus perform all the imports it needs.
+
+
+def _tk_test(count, func=None):
+    """
+    Run *func* in a subprocess, and assert it prints "success" *count* times.
+    """
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_backend_tk.py:21]:
```suggestion
def _isolated_tk_test(count, func=None):
    """
    A decorator to run *func* in a subprocess, and assert it prints "success"
    *success_count* times.

    TkAgg tests seem to have interactions between tests, so isolate each test
    in a subprocess. See GH#18261
    
    The decorated function must be fully self-contained, and thus perform
    all the imports it needs.
    """
```

And delete the comments above.

Comment (anntzer) [reply lib/matplotlib/tests/test_backend_tk.py]: yup

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/tests/test_backend_tk.py [hunk 2]
```diff
@@ -1,236 +1,181 @@
...
+        except subprocess.CalledProcessError:
+            pytest.fail("Subprocess failed to test intended behavior")
+        else:
+            assert proc.stdout.count("success") == success_count
+            # macOS may actually emit irrelevant errors about Accelerated
+            # OpenGL vs. software OpenGL, so suppress them.
+            assert not [line for line in proc.stderr.splitlines()
+                        if "OpenGL" not in line]
+
+    test_func.__name__ = func.__name__
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_tk.py:55]: Why not use `functools.wraps`?

File: lib/matplotlib/tests/test_backend_tk.py [hunk 3]
```diff
@@ -1,236 +1,181 @@
...
-    data = np.asarray(aggimage)
+@_isolated_tk_test(success_count=6)  # len(bad_boxes)
+def test_blit():  # pragma: no cover
+    import matplotlib.pyplot as plt
+    import numpy as np
+    from matplotlib.backends import _tkagg
+
+    fig, ax = plt.subplots()
+    photoimage = fig.canvas._tkphoto
+    data = np.asarray(np.ones((4, 4, 4)))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_tk.py:68]:
```suggestion
    data = np.ones((4, 4, 4))
```

File: lib/matplotlib/tests/test_backend_tk.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/tests/test_backend_tk.py:55]: I wasn't really sure if that semantically made sense as "wrapping" but I guess it does.

File: lib/matplotlib/tests/test_backend_tk.py (hunk 3 above)
Comment (anntzer) [reply lib/matplotlib/tests/test_backend_tk.py:68]: yes

File: lib/matplotlib/tests/test_backend_tk.py [hunk 4]
```diff
@@ -1,236 +1,181 @@
...
+        except subprocess.TimeoutExpired:
+            pytest.fail("Subprocess timed out")
+        except subprocess.CalledProcessError:
+            pytest.fail("Subprocess failed to test intended behavior")
+        else:
+            assert proc.stdout.count("success") == success_count
+            # macOS may actually emit irrelevant errors about Accelerated
+            # OpenGL vs. software OpenGL, so suppress them.
+            assert not [line for line in proc.stderr.splitlines()
+                        if "OpenGL" not in line]
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/tests/test_backend_tk.py]: I swapped the asserts.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**